# Stone Data Challenge 2022 - Scripts

O intuito deste script é apresentar o processo de preparação, transformação, processamento e análise dos dados para encontrar a solução ao problema **"Qual é a curva ideal de vezes que devemos acionar um cliente?"**.

## Preparação do ambiente

In [1]:
# Importando as bibliotecas necessárias

import pandas as pd
import numpy as np
import datetime as dt

In [2]:
# Arquivos carregados no Google Drive
# Permitindo que o Google Colab utilize os arquivos do Google Drive

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## Importação dos dados

In [3]:
# Definindo os caminhos dos arquivos CSV 

url_portfolio_clientes = '/content/drive/MyDrive/STONE Data Challenge 2022/portfolio_clientes.csv'
url_portfolio_tpv = '/content/drive/MyDrive/STONE Data Challenge 2022/portfolio_tpv.csv'
url_portfolio_comunicados = '/content/drive/MyDrive/STONE Data Challenge 2022/portfolio_comunicados.csv'
url_portfolio_geral = '/content/drive/MyDrive/STONE Data Challenge 2022/portfolio_geral.csv'

# Carregando os dados

dados_portfolio_clientes = pd.read_csv(url_portfolio_clientes)
dados_portfolio_tpv = pd.read_csv(url_portfolio_tpv)
dados_portfolio_comunicados = pd.read_csv(url_portfolio_comunicados)
dados_portfolio_geral = pd.read_csv(url_portfolio_geral)

In [ ]:
# Configurando para que os pandas apresente todas as colunas disponíveis

pd.set_option("display.max_columns", 22)

In [ ]:
# Verificando as primeiras linhas de portfolio_clientes

dados_portfolio_clientes.head()

nr_documento tipo_empresa                 cidade  \
0  094aaed740bbd7fcddb6c678c18d883a           PJ                Caucaia   
1  eaade7915611fcdcbba747e3a83365c6           PF       Cidade Ocidental   
2  1f21958fd296f1b04a1f0c2658913856          MEI                 Manaus   
3  fcdfe15db4520db1a1fe8b6a72878655          MEI  São Bernardo do Campo   
4  ba57e8837452b5e0114d7e273de039b4           PJ          Foz do Iguaçu   

  estado         subsegmento               segmento  
0     CE  Alimentação Rápida            Alimentação  
1     GO      Lojas Diversas                 Varejo  
2     AM  Alimentação Rápida            Alimentação  
3     SP  Alimentação Rápida            Alimentação  
4     PR      Lojas Diversas  Supermercado/Farmácia

In [ ]:
# Verificando as primeiras linhas de portfolio_tpv

dados_portfolio_tpv.head()

nr_documento  dt_transacao  qtd_transacoes  vlr_tpv
0  7f0464355620bc7c137b3546eb90867a      20210505               4    116.5
1  7f0464355620bc7c137b3546eb90867a      20210506              20    537.0
2  7f0464355620bc7c137b3546eb90867a      20210507              17    440.0
3  7f0464355620bc7c137b3546eb90867a      20210508              29   1328.0
4  7f0464355620bc7c137b3546eb90867a      20210509              35    969.5

In [ ]:
# Verificando as primeiras linhas de portfolio_comunicados

dados_portfolio_comunicados.head()

contrato_id dt_ref_portfolio   data_acao tipo_acao  \
0  0d823ddff1839eb938dccc68586ae365       2021-01-16  2021-01-18       HSM   
1  c2f8a21671dbaeaa4157e03682c639e1       2021-05-22  2021-05-24       HSM   
2  e19a222c233d46ff4984489c13aa1b98       2020-09-15  2020-09-15       HSM   
3  c24fc7630fe5b89928ba9478c2cdd9b2       2021-07-20  2021-07-20       HSM   
4  0abe9c2f21e941f2bd3b086277e33ebc       2020-12-26  2020-12-28       HSM   

                 acao        status  
0  campanhaobservacao          LIDO  
1  campanhaobservacao      ENTREGUE  
2  campanhaobservacao  NAO ENTREGUE  
3  campanhaobservacao  NAO ENTREGUE  
4  campanhaobservacao      ENTREGUE

In [ ]:
# Verificando as primeiras linhas de portfolio_geral

dados_portfolio_geral.head()

contrato_id dt_ref_portfolio    safra  \
0  356c02706c8e74b15004bb5964ade6bb       2022-03-01  2020-06   
1  1eb0e13f83f336146227ccc9efb08c47       2022-03-01  2020-07   
2  8b03675f597974ca96f06648e96af6d4       2022-03-01  2020-08   
3  4207c25cdd13aba5a05ae348e889d798       2022-03-01  2021-03   
4  7dfdca96b6ff0dd844e5e130aee88b30       2022-03-01  2020-11   

                       nr_documento status_contrato dt_contrato dt_desembolso  \
0  b0871d4e0d72afd1d44381e5b2453cb2         Settled  2020-06-24    2020-06-29   
1  938513461b4a4fcd2bf99ffd8f12b9be         Settled  2020-07-26    2020-07-28   
2  4303bd71ee3fe80aba4efc8dbd6c7d70         Settled  2020-08-13    2020-08-17   
3  9c1dc4403f8a4833bd58eea3ee85c917          Active  2021-03-01    2021-03-03   
4  3b6951df3075979c99a4dfd9f93efd38          Active  2020-10-31    2020-11-04   

  dt_vencimento       dt_wo  prazo  vlr_desembolsado  vlr_tarifa  juros_mes  \
0    2020-12-24  2021-12-24  10.26          22686.57         0.0    0.10788   
1    2021-10-26  2022-10-26  25.65          16074.00         0.0    0.04788   
2    2022-02-13  2023-02-13  30.78          17100.00         0.0    0.05388   
3    2022-06-01  2023-06-01  25.65          14588.01         0.0    0.04788   
4    2022-01-31  2023-01-31  25.65          68400.00         0.0    0.04788   

   juros_diario  perc_retencao  vlr_pgto_realizado  vlr_pgto_esperado  \
0       0.00345        0.24000                 0.0             163.00   
1       0.00157        0.08400                 0.0              46.87   
2       0.00176        0.08076                 0.0              45.57   
3       0.00157        0.19200                 0.0              43.06   
4       0.00157        1.20000                 0.0             200.09   

   vlr_saldo_devedor  vlr_saldo_devedor_esperado  dsp  dspp  flag_transacao  
0               0.55                        0.00    0     0               0  
1               0.00                        0.00    0     0               0  
2               0.00                        0.00    0     0               0  
3            9469.47                      461.96   20    82               0  
4           57783.55                        0.00  134   167               0

## Análise iniciais (na estrutura dos dados)

In [ ]:
# Informações gerais da tabela portfolio_clientes

dados_portfolio_clientes.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14405 entries, 0 to 14404
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   nr_documento  14405 non-null  object
 1   tipo_empresa  14405 non-null  object
 2   cidade        14405 non-null  object
 3   estado        14405 non-null  object
 4   subsegmento   14405 non-null  object
 5   segmento      14405 non-null  object
dtypes: object(6)
memory usage: 675.4+ KB


In [ ]:
# Informações gerais da tabela portfolio_tpv

dados_portfolio_tpv.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4408597 entries, 0 to 4408596
Data columns (total 4 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   nr_documento    4408597 non-null  object 
 1   dt_transacao    4408597 non-null  int64  
 2   qtd_transacoes  4408597 non-null  int64  
 3   vlr_tpv         4408597 non-null  float64
dtypes: float64(1), int64(2), object(1)
memory usage: 134.5+ MB


In [ ]:
# Informações gerais da tabela portfolio_comunicados

dados_portfolio_comunicados.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 403740 entries, 0 to 403739
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   contrato_id       403740 non-null  object
 1   dt_ref_portfolio  403740 non-null  object
 2   data_acao         403740 non-null  object
 3   tipo_acao         403740 non-null  object
 4   acao              403740 non-null  object
 5   status            403740 non-null  object
dtypes: object(6)
memory usage: 18.5+ MB


In [ ]:
# Informações gerais da tabela portfolio_geral

dados_portfolio_geral.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8061006 entries, 0 to 8061005
Data columns (total 22 columns):
 #   Column                      Non-Null Count    Dtype  
---  ------                      --------------    -----  
 0   contrato_id                 8061006 non-null  object 
 1   dt_ref_portfolio            8061006 non-null  object 
 2   safra                       8061006 non-null  object 
 3   nr_documento                8061006 non-null  object 
 4   status_contrato             8061006 non-null  object 
 5   dt_contrato                 8061006 non-null  object 
 6   dt_desembolso               8061006 non-null  object 
 7   dt_vencimento               8061006 non-null  object 
 8   dt_wo                       8061006 non-null  object 
 9   prazo                       8061006 non-null  float64
 10  vlr_desembolsado            8061006 non-null  float64
 11  vlr_tarifa                  8061006 non-null  float64
 12  juros_mes                   8061006 non-null  float64
 1

*   Verifiquei que não há dados faltantes.
*   Verifiquei que a variável “dt_transacao” da tabela “portfolio_tpv” deveria ser do tipo data e está como tipo texto. Mas no Python foi identificada como tipo inteiro (int64).



In [ ]:
# Verificando dados duplicados na tabela portfolio_clientes 
# onde a variável a ser analisada é a variável 'nr_documento', chave primária da tabela

dados_portfolio_clientes[dados_portfolio_clientes.duplicated(keep = False,
                                                             subset = ['nr_documento'])].sort_values(by = 'nr_documento')

nr_documento tipo_empresa          cidade estado  \
7776   0044c33f545a655f12097beefc124350           PJ           Belém     PA   
7823   0044c33f545a655f12097beefc124350           PJ           Belém     PA   
5612   00fbc2dd1bde4a53f4fed5ff33c8f26c           PJ   Novo Hamburgo     RS   
11839  00fbc2dd1bde4a53f4fed5ff33c8f26c           PJ   Novo Hamburgo     RS   
6239   010d8a4b8ee4fac79794a941f60805b5           PJ       Joinville     SC   
...                                 ...          ...             ...    ...   
1396   fd4af637f5961e3c402e6ac7e7a6b096           PF          Olinda     PE   
13852  fdade62eb335a3dd3256c986f8120394           PF  Rio de Janeiro     RJ   
9996   fdade62eb335a3dd3256c986f8120394           PJ  Rio de Janeiro     RJ   
4909   fecc9d3f0ee84352963107be7a5726e4           PJ       São Paulo     SP   
4905   fecc9d3f0ee84352963107be7a5726e4           PJ       São Paulo     SP   

                  subsegmento               segmento  
7776                     None                 Outros  
7823   Material de Construção          Bens duráveis  
5612                     None                 Outros  
11839              Automotivo          Bens duráveis  
6239                Vestuário                 Varejo  
...                       ...                    ...  
1396           Lojas Diversas                 Varejo  
13852          Lojas Diversas                 Varejo  
9996                    Saúde  Supermercado/Farmácia  
4909                    Saúde  Supermercado/Farmácia  
4905                    Saúde  Supermercado/Farmácia  

[280 rows x 6 columns]

Verifiquei que a tabela “portfolio_clientes” possuem dados duplicados na variável “nr_documento. Por ser uma tabela de dados cadastrais de clientes, ela deve possuir valores únicos na variável “nr_documento”, que é a chave primária dessa tabela.

In [ ]:
# Verificando dados duplicados na tabela portfolio_tpv 
# onde as variáveis a serem analisadas são: 'nr_documento' e 'dt_transacao', chaves primárias da tabela

dados_portfolio_tpv[dados_portfolio_tpv.duplicated(keep = False,
                                                   subset = ['nr_documento', 'dt_transacao'])].sort_values(by = 'nr_documento')

Empty DataFrame
Columns: [nr_documento, dt_transacao, qtd_transacoes, vlr_tpv]
Index: []

In [ ]:
dados_portfolio_tpv[dados_portfolio_tpv.duplicated(keep = False)]

Empty DataFrame
Columns: [nr_documento, dt_transacao, qtd_transacoes, vlr_tpv]
Index: []

Verifiquei que a tabela “portfolio_tpv” não possui dados duplicados.

In [ ]:
# Verificando dados duplicados na tabela portfolio_comunicados 
# onde as variáveis a serem analisadas são: 'contrato_id', 'dt_ref_portfolio' e 'tipo_acao', chaves primárias da tabela

# Para análise dos valores duplicados, desconsiderei as comunicações com status = 'NAO ENTREGUE'

dados_entregues = dados_portfolio_comunicados[dados_portfolio_comunicados['status'] != 'NAO ENTREGUE']

dados_duplicados = dados_entregues[dados_entregues.duplicated(keep = False,
                                                          subset = ['contrato_id', 'dt_ref_portfolio', 'tipo_acao'])]
dados_duplicados.sort_values(by = ['contrato_id', 'dt_ref_portfolio', 'tipo_acao'])

contrato_id dt_ref_portfolio   data_acao  \
49842   000c35a61297edadc2842f6d5b4028e1       2021-05-09  2021-05-11   
206361  000c35a61297edadc2842f6d5b4028e1       2021-05-09  2021-05-11   
274715  000dcdc93a545ee45a1aee85ef85c34a       2020-11-10  2020-11-10   
403563  000dcdc93a545ee45a1aee85ef85c34a       2020-11-10  2020-11-10   
48002   000dcdc93a545ee45a1aee85ef85c34a       2021-01-30  2021-02-01   
...                                  ...              ...         ...   
331727  ffedadd25ac72b41676b32a79adc0333       2022-04-14  2022-04-14   
22365   ffedadd25ac72b41676b32a79adc0333       2022-04-14  2022-04-14   
354875  ffedadd25ac72b41676b32a79adc0333       2022-04-14  2022-04-14   
154766  fffb0d6ec51d163ca8c69900f748d61b       2022-03-01  2022-03-01   
347841  fffb0d6ec51d163ca8c69900f748d61b       2022-03-01  2022-03-01   

       tipo_acao                    acao    status  
49842        HSM  campanhaprenegativacao      LIDO  
206361       HSM    campanhaparcelamento  ENTREGUE  
274715       HSM      campanhaobservacao      LIDO  
403563       HSM   campanhaboletoquitado  ENTREGUE  
48002      EMAIL  campanhaprenegativacao  ENTREGUE  
...          ...                     ...       ...  
331727     EMAIL    campanhaparcelamento  ENTREGUE  
22365        HSM    campanhaparcelamento  ENTREGUE  
354875       HSM  campanhaprenegativacao  ENTREGUE  
154766     EMAIL      campanhaobservacao      LIDO  
347841     EMAIL   campanhaboletoquitado  ENTREGUE  

[12506 rows x 6 columns]

In [ ]:
# Deletando as variaveis criadas apenas para análise

del(dados_entregues)
del(dados_duplicados)

In [ ]:
dados_portfolio_comunicados[dados_portfolio_comunicados.duplicated(keep = False)]

Empty DataFrame
Columns: [contrato_id, dt_ref_portfolio, data_acao, tipo_acao, acao, status]
Index: []

Na tabela “portfolio_comunicados”, verifiquei que para os registros que possuem os mesmos valores nas variáveis “contrato_id”, “dt_ref_portfolio”, “data_acao”, e “tipo_acao” existem campanhas diferentes de comunicação. Ou seja, para o mesmo contrato e no mesmo dia foram enviadas mais de uma comunicação (campanha). Estou assumindo que isso acontece por causa dos momentos diferentes da jornada do contrato do cliente (como realçado na descrição do case).

In [ ]:
# Verificando dados duplicados na tabela portfolio_geral 
# onde as variáveis a serem analisadas são: 
# 'contrato_id' e 'dt_ref_portfolio' (chaves primárias da tabela) e 'nr_documento' (chave estrangeira da tabela)

dados_portfolio_geral[dados_portfolio_geral.duplicated(keep = False,
                                                       subset = ['contrato_id', 'dt_ref_portfolio'])].sort_values(by = 'contrato_id')

Empty DataFrame
Columns: [contrato_id, dt_ref_portfolio, safra, nr_documento, status_contrato, dt_contrato, dt_desembolso, dt_vencimento, dt_wo, prazo, vlr_desembolsado, vlr_tarifa, juros_mes, juros_diario, perc_retencao, vlr_pgto_realizado, vlr_pgto_esperado, vlr_saldo_devedor, vlr_saldo_devedor_esperado, dsp, dspp, flag_transacao]
Index: []

In [ ]:
dados_portfolio_geral[dados_portfolio_geral.duplicated(keep = False)]

Empty DataFrame
Columns: [contrato_id, dt_ref_portfolio, safra, nr_documento, status_contrato, dt_contrato, dt_desembolso, dt_vencimento, dt_wo, prazo, vlr_desembolsado, vlr_tarifa, juros_mes, juros_diario, perc_retencao, vlr_pgto_realizado, vlr_pgto_esperado, vlr_saldo_devedor, vlr_saldo_devedor_esperado, dsp, dspp, flag_transacao]
Index: []

Verifiquei que a tabela “portfolio_geral” não possui dados duplicados.

## Preparação e Limpeza dos dados

In [4]:
# Corrigindo o formato da data da variável dt_transacao da tabela portfolio_tpv

# Definindo função para formatar a data no formato aaaa-mm-dd
def formata_data(x):
  x = str(x)
  ano = x[0:4]
  mes = x[4:6]
  dia = x[6:8]  
  return ano + '-' + mes + '-' + dia

# Transformando a data "dt_transacao" no formato aaaa-mm-dd
dados_portfolio_tpv['dt_transacao'] = dados_portfolio_tpv['dt_transacao'].apply(formata_data)

In [ ]:
dados_portfolio_tpv.head()

nr_documento dt_transacao  qtd_transacoes  vlr_tpv
0  7f0464355620bc7c137b3546eb90867a   2021-05-05               4    116.5
1  7f0464355620bc7c137b3546eb90867a   2021-05-06              20    537.0
2  7f0464355620bc7c137b3546eb90867a   2021-05-07              17    440.0
3  7f0464355620bc7c137b3546eb90867a   2021-05-08              29   1328.0
4  7f0464355620bc7c137b3546eb90867a   2021-05-09              35    969.5

In [ ]:
dados_portfolio_tpv.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4408597 entries, 0 to 4408596
Data columns (total 4 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   nr_documento    4408597 non-null  object 
 1   dt_transacao    4408597 non-null  object 
 2   qtd_transacoes  4408597 non-null  int64  
 3   vlr_tpv         4408597 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 134.5+ MB


In [5]:
# Removendo dados duplicados da tabela portfolio_clientes 

dados_portfolio_clientes.drop_duplicates(keep = 'last',
                                         subset = ['nr_documento'],
                                         inplace = True)
dados_portfolio_clientes.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14265 entries, 0 to 14404
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   nr_documento  14265 non-null  object
 1   tipo_empresa  14265 non-null  object
 2   cidade        14265 non-null  object
 3   estado        14265 non-null  object
 4   subsegmento   14265 non-null  object
 5   segmento      14265 non-null  object
dtypes: object(6)
memory usage: 780.1+ KB


## Transformação das tabelas em uma tabela

In [6]:
# Filtrando a tabela portfolio_comunicados para desconsiderar os comunicados não entregues 

dados_portfolio_comunicados_filtrado = dados_portfolio_comunicados[dados_portfolio_comunicados['status'] != 'NAO ENTREGUE']
dados_portfolio_comunicados_filtrado.head()

contrato_id dt_ref_portfolio   data_acao tipo_acao  \
0  0d823ddff1839eb938dccc68586ae365       2021-01-16  2021-01-18       HSM   
1  c2f8a21671dbaeaa4157e03682c639e1       2021-05-22  2021-05-24       HSM   
4  0abe9c2f21e941f2bd3b086277e33ebc       2020-12-26  2020-12-28       HSM   
5  90329f6c65f6b7f07e4fd251925a90b9       2020-05-16  2020-05-18       HSM   
6  b4c216d4e06517e219528d21a4d8b28a       2021-09-11  2021-09-13       HSM   

                 acao    status  
0  campanhaobservacao      LIDO  
1  campanhaobservacao  ENTREGUE  
4  campanhaobservacao  ENTREGUE  
5  campanhaobservacao  ENTREGUE  
6  campanhaobservacao  ENTREGUE

In [ ]:
dados_portfolio_comunicados_filtrado.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 212533 entries, 0 to 403736
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   contrato_id       212533 non-null  object
 1   dt_ref_portfolio  212533 non-null  object
 2   data_acao         212533 non-null  object
 3   tipo_acao         212533 non-null  object
 4   acao              212533 non-null  object
 5   status            212533 non-null  object
dtypes: object(6)
memory usage: 11.4+ MB


In [7]:
# Selecionando as variáveis da tabela portfolio_geral

dados_portfolio_geral = dados_portfolio_geral[['contrato_id', 'dt_ref_portfolio', 'nr_documento',  
                                               'vlr_pgto_esperado', 'vlr_pgto_realizado', 'vlr_saldo_devedor', 'vlr_saldo_devedor_esperado', 
                                               'vlr_desembolsado', 'status_contrato', 
                                               'dsp', 'dspp', 'flag_transacao']]
dados_portfolio_geral.head()

contrato_id dt_ref_portfolio  \
0  356c02706c8e74b15004bb5964ade6bb       2022-03-01   
1  1eb0e13f83f336146227ccc9efb08c47       2022-03-01   
2  8b03675f597974ca96f06648e96af6d4       2022-03-01   
3  4207c25cdd13aba5a05ae348e889d798       2022-03-01   
4  7dfdca96b6ff0dd844e5e130aee88b30       2022-03-01   

                       nr_documento  vlr_pgto_esperado  vlr_pgto_realizado  \
0  b0871d4e0d72afd1d44381e5b2453cb2             163.00                 0.0   
1  938513461b4a4fcd2bf99ffd8f12b9be              46.87                 0.0   
2  4303bd71ee3fe80aba4efc8dbd6c7d70              45.57                 0.0   
3  9c1dc4403f8a4833bd58eea3ee85c917              43.06                 0.0   
4  3b6951df3075979c99a4dfd9f93efd38             200.09                 0.0   

   vlr_saldo_devedor  vlr_saldo_devedor_esperado  vlr_desembolsado  \
0               0.55                        0.00          22686.57   
1               0.00                        0.00          16074.00   
2               0.00                        0.00          17100.00   
3            9469.47                      461.96          14588.01   
4           57783.55                        0.00          68400.00   

  status_contrato  dsp  dspp  flag_transacao  
0         Settled    0     0               0  
1         Settled    0     0               0  
2         Settled    0     0               0  
3          Active   20    82               0  
4          Active  134   167               0

In [ ]:
dados_portfolio_geral.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8061006 entries, 0 to 8061005
Data columns (total 12 columns):
 #   Column                      Non-Null Count    Dtype  
---  ------                      --------------    -----  
 0   contrato_id                 8061006 non-null  object 
 1   dt_ref_portfolio            8061006 non-null  object 
 2   nr_documento                8061006 non-null  object 
 3   vlr_pgto_esperado           8061006 non-null  float64
 4   vlr_pgto_realizado          8061006 non-null  float64
 5   vlr_saldo_devedor           8061006 non-null  float64
 6   vlr_saldo_devedor_esperado  8061006 non-null  float64
 7   vlr_desembolsado            8061006 non-null  float64
 8   status_contrato             8061006 non-null  object 
 9   dsp                         8061006 non-null  int64  
 10  dspp                        8061006 non-null  int64  
 11  flag_transacao              8061006 non-null  int64  
dtypes: float64(5), int64(3), object(4)
memory usage: 738.0+ 

In [8]:
# Criando a tabela única (portfolio) a partir da tabela portfolio_comunicados
# E unindo-a com a tabela portfolio_geral

dados_portfolio = pd.merge( 
    left = dados_portfolio_comunicados_filtrado, 
    right = dados_portfolio_geral, 
    left_on = ['contrato_id', 'dt_ref_portfolio'], 
    right_on = ['contrato_id', 'dt_ref_portfolio'], 
    how = 'left'
)
dados_portfolio.head(5) 

contrato_id dt_ref_portfolio   data_acao tipo_acao  \
0  0d823ddff1839eb938dccc68586ae365       2021-01-16  2021-01-18       HSM   
1  c2f8a21671dbaeaa4157e03682c639e1       2021-05-22  2021-05-24       HSM   
2  0abe9c2f21e941f2bd3b086277e33ebc       2020-12-26  2020-12-28       HSM   
3  90329f6c65f6b7f07e4fd251925a90b9       2020-05-16  2020-05-18       HSM   
4  b4c216d4e06517e219528d21a4d8b28a       2021-09-11  2021-09-13       HSM   

                 acao    status                      nr_documento  \
0  campanhaobservacao      LIDO  f3514fae3c00ff2b5f8b3b0ded1c4939   
1  campanhaobservacao  ENTREGUE  f4e083d0e86f07a7c0f9485b6c4e09b4   
2  campanhaobservacao  ENTREGUE  cb29bfb93cd1a62c11a27d2c17771ba2   
3  campanhaobservacao  ENTREGUE  4902ae1a11b3817caa938e750c6a1ce3   
4  campanhaobservacao  ENTREGUE  2cd211b42e3cfb8ce0aae537daec2ce4   

   vlr_pgto_esperado  vlr_pgto_realizado  vlr_saldo_devedor  \
0              35.64                 0.0           10430.74   
1              29.39                 0.0            5963.78   
2             167.83                 0.0           47701.51   
3              76.21                 0.0           22034.68   
4              78.44                 0.0           14462.33   

   vlr_saldo_devedor_esperado  vlr_desembolsado status_contrato  dsp  dspp  \
0                     9407.99          11439.90          Active    5    68   
1                     5746.55           5843.07          Active    5     5   
2                    43630.81          53873.55          Active    2    15   
3                    21247.21          22713.93          Active    5     5   
4                     9043.51          26812.80          Active    5     5   

   flag_transacao  
0               1  
1               1  
2               0  
3               1  
4               1

In [ ]:
dados_portfolio.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 212533 entries, 0 to 212532
Data columns (total 16 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   contrato_id                 212533 non-null  object 
 1   dt_ref_portfolio            212533 non-null  object 
 2   data_acao                   212533 non-null  object 
 3   tipo_acao                   212533 non-null  object 
 4   acao                        212533 non-null  object 
 5   status                      212533 non-null  object 
 6   nr_documento                212533 non-null  object 
 7   vlr_pgto_esperado           212533 non-null  float64
 8   vlr_pgto_realizado          212533 non-null  float64
 9   vlr_saldo_devedor           212533 non-null  float64
 10  vlr_saldo_devedor_esperado  212533 non-null  float64
 11  vlr_desembolsado            212533 non-null  float64
 12  status_contrato             212533 non-null  object 
 13  dsp           

In [9]:
# Combinando a tabela única (portfolio) e a tabela portfolio_clientes

dados_portfolio = pd.merge( 
    left = dados_portfolio, 
    right = dados_portfolio_clientes, 
    left_on = 'nr_documento', 
    right_on = 'nr_documento', 
    how = 'left'
)
dados_portfolio.head(5) 

contrato_id dt_ref_portfolio   data_acao tipo_acao  \
0  0d823ddff1839eb938dccc68586ae365       2021-01-16  2021-01-18       HSM   
1  c2f8a21671dbaeaa4157e03682c639e1       2021-05-22  2021-05-24       HSM   
2  0abe9c2f21e941f2bd3b086277e33ebc       2020-12-26  2020-12-28       HSM   
3  90329f6c65f6b7f07e4fd251925a90b9       2020-05-16  2020-05-18       HSM   
4  b4c216d4e06517e219528d21a4d8b28a       2021-09-11  2021-09-13       HSM   

                 acao    status                      nr_documento  \
0  campanhaobservacao      LIDO  f3514fae3c00ff2b5f8b3b0ded1c4939   
1  campanhaobservacao  ENTREGUE  f4e083d0e86f07a7c0f9485b6c4e09b4   
2  campanhaobservacao  ENTREGUE  cb29bfb93cd1a62c11a27d2c17771ba2   
3  campanhaobservacao  ENTREGUE  4902ae1a11b3817caa938e750c6a1ce3   
4  campanhaobservacao  ENTREGUE  2cd211b42e3cfb8ce0aae537daec2ce4   

   vlr_pgto_esperado  vlr_pgto_realizado  vlr_saldo_devedor  ...  \
0              35.64                 0.0           10430.74  ...   
1              29.39                 0.0            5963.78  ...   
2             167.83                 0.0           47701.51  ...   
3              76.21                 0.0           22034.68  ...   
4              78.44                 0.0           14462.33  ...   

   vlr_desembolsado  status_contrato dsp  dspp  flag_transacao  tipo_empresa  \
0          11439.90           Active   5    68               1           MEI   
1           5843.07           Active   5     5               1            PF   
2          53873.55           Active   2    15               0           MEI   
3          22713.93           Active   5     5               1           MEI   
4          26812.80           Active   5     5               1            PF   

                    cidade estado             subsegmento  \
0               Petrópolis     RJ               Vestuário   
1  Jaboatao Dos Guararapes     PE      Alimentação Rápida   
2                 Salvador     BA                  Outros   
3                São Paulo     SP    Bares e Restaurantes   
4                   Manaus     AM  Material de Construção   

                   segmento  
0                    Varejo  
1               Alimentação  
2                  Serviços  
3  Viagens e entretenimento  
4             Bens duráveis  

[5 rows x 21 columns]

In [ ]:
dados_portfolio.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 212533 entries, 0 to 212532
Data columns (total 21 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   contrato_id                 212533 non-null  object 
 1   dt_ref_portfolio            212533 non-null  object 
 2   data_acao                   212533 non-null  object 
 3   tipo_acao                   212533 non-null  object 
 4   acao                        212533 non-null  object 
 5   status                      212533 non-null  object 
 6   nr_documento                212533 non-null  object 
 7   vlr_pgto_esperado           212533 non-null  float64
 8   vlr_pgto_realizado          212533 non-null  float64
 9   vlr_saldo_devedor           212533 non-null  float64
 10  vlr_saldo_devedor_esperado  212533 non-null  float64
 11  vlr_desembolsado            212533 non-null  float64
 12  status_contrato             212533 non-null  object 
 13  dsp           

In [10]:
# Combinando a tabela única (portfolio) e a tabela portfolio_tpv

dados_portfolio = pd.merge( 
    left = dados_portfolio, 
    right = dados_portfolio_tpv, 
    left_on = ['nr_documento', 'dt_ref_portfolio'], 
    right_on = ['nr_documento', 'dt_transacao'], 
    how = 'left'
)
dados_portfolio.head(5) 

contrato_id dt_ref_portfolio   data_acao tipo_acao  \
0  0d823ddff1839eb938dccc68586ae365       2021-01-16  2021-01-18       HSM   
1  c2f8a21671dbaeaa4157e03682c639e1       2021-05-22  2021-05-24       HSM   
2  0abe9c2f21e941f2bd3b086277e33ebc       2020-12-26  2020-12-28       HSM   
3  90329f6c65f6b7f07e4fd251925a90b9       2020-05-16  2020-05-18       HSM   
4  b4c216d4e06517e219528d21a4d8b28a       2021-09-11  2021-09-13       HSM   

                 acao    status                      nr_documento  \
0  campanhaobservacao      LIDO  f3514fae3c00ff2b5f8b3b0ded1c4939   
1  campanhaobservacao  ENTREGUE  f4e083d0e86f07a7c0f9485b6c4e09b4   
2  campanhaobservacao  ENTREGUE  cb29bfb93cd1a62c11a27d2c17771ba2   
3  campanhaobservacao  ENTREGUE  4902ae1a11b3817caa938e750c6a1ce3   
4  campanhaobservacao  ENTREGUE  2cd211b42e3cfb8ce0aae537daec2ce4   

   vlr_pgto_esperado  vlr_pgto_realizado  vlr_saldo_devedor  ...  dspp  \
0              35.64                 0.0           10430.74  ...    68   
1              29.39                 0.0            5963.78  ...     5   
2             167.83                 0.0           47701.51  ...    15   
3              76.21                 0.0           22034.68  ...     5   
4              78.44                 0.0           14462.33  ...     5   

   flag_transacao tipo_empresa                   cidade  estado  \
0               1          MEI               Petrópolis      RJ   
1               1           PF  Jaboatao Dos Guararapes      PE   
2               0          MEI                 Salvador      BA   
3               1          MEI                São Paulo      SP   
4               1           PF                   Manaus      AM   

              subsegmento                  segmento dt_transacao  \
0               Vestuário                    Varejo   2021-01-16   
1      Alimentação Rápida               Alimentação   2021-05-22   
2                  Outros                  Serviços          NaN   
3    Bares e Restaurantes  Viagens e entretenimento   2020-05-16   
4  Material de Construção             Bens duráveis          NaN   

  qtd_transacoes vlr_tpv  
0            4.0   359.6  
1            7.0   223.5  
2            NaN     NaN  
3            9.0   345.0  
4            NaN     NaN  

[5 rows x 24 columns]

In [ ]:
dados_portfolio.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 212533 entries, 0 to 212532
Data columns (total 24 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   contrato_id                 212533 non-null  object 
 1   dt_ref_portfolio            212533 non-null  object 
 2   data_acao                   212533 non-null  object 
 3   tipo_acao                   212533 non-null  object 
 4   acao                        212533 non-null  object 
 5   status                      212533 non-null  object 
 6   nr_documento                212533 non-null  object 
 7   vlr_pgto_esperado           212533 non-null  float64
 8   vlr_pgto_realizado          212533 non-null  float64
 9   vlr_saldo_devedor           212533 non-null  float64
 10  vlr_saldo_devedor_esperado  212533 non-null  float64
 11  vlr_desembolsado            212533 non-null  float64
 12  status_contrato             212533 non-null  object 
 13  dsp           

## Criação de Novas Variáveis

In [11]:
# Criando a variável 'semana_acao' a partir da variável 'data_acao' da tabela final portfolio

# Lista com os dias da semana
dias_semana = ["Segunda", "Terça", "Quarta", "Quinta", "Sexta", "Sábado", "Domingo"]

# Função para transformar o dia da semana de número para nome a partir de uma data no formato 'aaaa-mm-dd'
def semana(data_texto):
  data = dt.datetime.strptime(data_texto, '%Y-%m-%d')
  return dias_semana[data.weekday()]

# Aplicando a função acima para criar variável 'semana_acao' na tabela portfolio
dados_portfolio['semana_acao'] = dados_portfolio['data_acao'].apply(semana)

In [12]:
dados_portfolio[['data_acao', 'semana_acao']]

data_acao semana_acao
0       2021-01-18     Segunda
1       2021-05-24     Segunda
2       2020-12-28     Segunda
3       2020-05-18     Segunda
4       2021-09-13     Segunda
...            ...         ...
212528  2021-01-20      Quarta
212529  2021-11-11      Quinta
212530  2020-10-20       Terça
212531  2022-04-05       Terça
212532  2020-06-19       Sexta

[212533 rows x 2 columns]

In [13]:
# Criando a variável 'ano_acao' a partir da variável 'data_acao' da tabela final portfolio

# Função para extrair o ano de uma data no formato 'aaaa-mm-dd'
def ano(data_texto):
  data = dt.datetime.strptime(data_texto, '%Y-%m-%d')
  return data.year

# Aplicando a função acima para criar variável 'ano_acao' na tabela portfolio
dados_portfolio['ano_acao'] = dados_portfolio['data_acao'].apply(ano)

In [14]:
dados_portfolio[['data_acao', 'ano_acao']]

data_acao  ano_acao
0       2021-01-18      2021
1       2021-05-24      2021
2       2020-12-28      2020
3       2020-05-18      2020
4       2021-09-13      2021
...            ...       ...
212528  2021-01-20      2021
212529  2021-11-11      2021
212530  2020-10-20      2020
212531  2022-04-05      2022
212532  2020-06-19      2020

[212533 rows x 2 columns]

In [15]:
# Criando a variável 'mes_abrev_acao' a partir da variável 'data_acao' da tabela final portfolio

# Lista com os meses abreviados
meses_abreviados = ["JAN", "FEV", "MAR", "ABR", "MAI", "JUN", "JUL", "AGO", "SET", "OUT", "NOV", "DEZ"]

# Função para transformar o mês em nome do mês abreviado a partir de uma data no formato 'aaaa-mm-dd'
def mes_abreviado(data_texto):
  data = dt.datetime.strptime(data_texto, '%Y-%m-%d')
  return meses_abreviados[data.month-1]

# Aplicando a função acima para criar variável 'mes_abrev_acao' na tabela portfolio
dados_portfolio['mes_abrev_acao'] = dados_portfolio['data_acao'].apply(mes_abreviado)

In [16]:
dados_portfolio[['data_acao', 'mes_abrev_acao']]

data_acao mes_abrev_acao
0       2021-01-18            JAN
1       2021-05-24            MAI
2       2020-12-28            DEZ
3       2020-05-18            MAI
4       2021-09-13            SET
...            ...            ...
212528  2021-01-20            JAN
212529  2021-11-11            NOV
212530  2020-10-20            OUT
212531  2022-04-05            ABR
212532  2020-06-19            JUN

[212533 rows x 2 columns]

In [17]:
# Verificando a tabela final com todas as variáveis

pd.set_option("display.max_columns", 27)
dados_portfolio.head()

contrato_id dt_ref_portfolio   data_acao tipo_acao  \
0  0d823ddff1839eb938dccc68586ae365       2021-01-16  2021-01-18       HSM   
1  c2f8a21671dbaeaa4157e03682c639e1       2021-05-22  2021-05-24       HSM   
2  0abe9c2f21e941f2bd3b086277e33ebc       2020-12-26  2020-12-28       HSM   
3  90329f6c65f6b7f07e4fd251925a90b9       2020-05-16  2020-05-18       HSM   
4  b4c216d4e06517e219528d21a4d8b28a       2021-09-11  2021-09-13       HSM   

                 acao    status                      nr_documento  \
0  campanhaobservacao      LIDO  f3514fae3c00ff2b5f8b3b0ded1c4939   
1  campanhaobservacao  ENTREGUE  f4e083d0e86f07a7c0f9485b6c4e09b4   
2  campanhaobservacao  ENTREGUE  cb29bfb93cd1a62c11a27d2c17771ba2   
3  campanhaobservacao  ENTREGUE  4902ae1a11b3817caa938e750c6a1ce3   
4  campanhaobservacao  ENTREGUE  2cd211b42e3cfb8ce0aae537daec2ce4   

   vlr_pgto_esperado  vlr_pgto_realizado  vlr_saldo_devedor  \
0              35.64                 0.0           10430.74   
1              29.39                 0.0            5963.78   
2             167.83                 0.0           47701.51   
3              76.21                 0.0           22034.68   
4              78.44                 0.0           14462.33   

   vlr_saldo_devedor_esperado  vlr_desembolsado status_contrato  dsp  dspp  \
0                     9407.99          11439.90          Active    5    68   
1                     5746.55           5843.07          Active    5     5   
2                    43630.81          53873.55          Active    2    15   
3                    21247.21          22713.93          Active    5     5   
4                     9043.51          26812.80          Active    5     5   

   flag_transacao tipo_empresa                   cidade estado  \
0               1          MEI               Petrópolis     RJ   
1               1           PF  Jaboatao Dos Guararapes     PE   
2               0          MEI                 Salvador     BA   
3               1          MEI                São Paulo     SP   
4               1           PF                   Manaus     AM   

              subsegmento                  segmento dt_transacao  \
0               Vestuário                    Varejo   2021-01-16   
1      Alimentação Rápida               Alimentação   2021-05-22   
2                  Outros                  Serviços          NaN   
3    Bares e Restaurantes  Viagens e entretenimento   2020-05-16   
4  Material de Construção             Bens duráveis          NaN   

   qtd_transacoes  vlr_tpv semana_acao  ano_acao mes_abrev_acao  
0             4.0    359.6     Segunda      2021            JAN  
1             7.0    223.5     Segunda      2021            MAI  
2             NaN      NaN     Segunda      2020            DEZ  
3             9.0    345.0     Segunda      2020            MAI  
4             NaN      NaN     Segunda      2021            SET

In [ ]:
dados_portfolio.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 212533 entries, 0 to 212532
Data columns (total 27 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   contrato_id                 212533 non-null  object 
 1   dt_ref_portfolio            212533 non-null  object 
 2   data_acao                   212533 non-null  object 
 3   tipo_acao                   212533 non-null  object 
 4   acao                        212533 non-null  object 
 5   status                      212533 non-null  object 
 6   nr_documento                212533 non-null  object 
 7   vlr_pgto_esperado           212533 non-null  float64
 8   vlr_pgto_realizado          212533 non-null  float64
 9   vlr_saldo_devedor           212533 non-null  float64
 10  vlr_saldo_devedor_esperado  212533 non-null  float64
 11  vlr_desembolsado            212533 non-null  float64
 12  status_contrato             212533 non-null  object 
 13  dsp           

## Exportação da tabela final

In [18]:
# Exportando a tabela única criada (portfolio) para o Google Drive

url = '/content/drive/MyDrive/STONE Data Challenge 2022/Tabela Final/Tabela_Final.csv'
dados_portfolio.to_csv(url)

## Validações iniciais

In [ ]:
# Qtde de contratos distintos na tabela final (portfolio)
# Ou seja, verificando a quantidade de contratos que receberem comunicados e que foram entregues, lidos ou respondidos

qtde_contratos_comunicados_entregues = len(np.unique(dados_portfolio['contrato_id']))
print(f'Contratos distintos na observação final: {qtde_contratos_comunicados_entregues}')

Contratos distintos na observação final: 11897


In [ ]:
# Analisando os filtros de exclusão dos contratos

qtde_contratos_comunicados = len(np.unique(dados_portfolio_comunicados['contrato_id']))

print(f'Quantidade de contratos que receberam algum comunicado: {qtde_contratos_comunicados}')
print(f"Quantidade de contratos com comunicados 'NAO ENTREGUE': {qtde_contratos_comunicados - qtde_contratos_comunicados_entregues}")

Quantidade de contratos que receberam algum comunicado: 12202
Quantidade de contratos com comunicados 'NAO ENTREGUE': 305


Logo, temos 305 contratos deveriam receber algum comunicado, mas, por algum motivo, o comunicado não foi entregue.